dali/pipeline/operator/sequence_operator.h

```diff
@@ -0,0 +1,627 @@
+                    "have exactly the same outermost dimensions to expand. However, got `",
+                    input_desc.ExpandedLayout(), "` planned for expanding."));
+    for (int sample_idx = 0; sample_idx < expand_desc.NumSamples(); ++sample_idx) {
+      if (expand_desc.ExpandFrames()) {
+        DALI_ENFORCE(expand_desc.NumFrames(sample_idx) == input_desc.NumFrames(sample_idx),
+                     make_string("Inputs ", expand_idx, " and ", input_idx,
+                                 " have different number of frames for sample ", sample_idx,
+                                 ", respectively: ", expand_desc.NumFrames(sample_idx), " and ",
+                                 input_desc.NumFrames(sample_idx)));
+      }
+      if (expand_desc.ExpandChannels()) {
+        DALI_ENFORCE(expand_desc.NumChannels(sample_idx) == input_desc.NumChannels(sample_idx),
+                     make_string("Inputs ", expand_idx, " and ", input_idx,
+                                 " have different number of channels for sample ", sample_idx,
+                                 ", respectively: ", expand_desc.NumChannels(sample_idx), " and ",
+                                 input_desc.NumChannels(sample_idx)));
+      }
+    }
+  }
+
+  void AddInputExpandedLike(const workspace_t<Backend> &ws, int input_idx,
+                            const ExpandDesc &input_desc) {
+    auto num_expand_dims = input_desc.NumDimsToExpand();
+    if (ws.template InputIsType<GPUBackend>(input_idx)) {
+      auto expanded_handle = ExpandInput<GPUBackend>(ws, input_idx, num_expand_dims);
+      AddInputHelper<GPUBackend>(expanded_handle);
+    } else {
+      auto expanded_handle = ExpandInput<CPUBackend>(ws, input_idx, num_expand_dims);
+      AddInputHelper<CPUBackend>(expanded_handle);
+    }
+  }
+
+  template <typename InputBackend>
+  ws_input_t<InputBackend> ExpandInput(const workspace_t<Backend> &ws, int input_idx,
+                                       int num_expand_dims) {
+    const auto &input = ws.template Input<InputBackend>(input_idx);
+    auto sample_dim = input.shape().sample_dim();
+    DALI_ENFORCE(
+        sample_dim > num_expand_dims,
+        make_string("Cannot flatten the sequence-like input ", input_idx,
+                    ". Samples must have more dimensions (got ", sample_dim,
+                    ") than the requested number of dimensions to unfold: ", num_expand_dims, "."));
+    // TODO(ktokarski) TODO(klecki)
+    // Rework it when TensorList stops being contigious and supports "true sample" mode
+    auto expanded_input = unfold_outer_dims(input, num_expand_dims);
+    const auto &input_layout = input.GetLayout();
+    expanded_input.SetLayout(input_layout.sub(num_expand_dims));
+    return std::make_shared<typename ws_input_t<InputBackend>::element_type>(
+        std::move(expanded_input));
+  }
+
+  template <typename OutputBackend>
+  ws_input_t<OutputBackend> ExpandOutput(const workspace_t<Backend> &ws, int output_idx,
```

Comment:
They are called UnfoldInput and UnfoldOutput now. To be honest I don't see much in common too - I'd like to have different error messages so that gives clearer error reporting and it is like half of the function.